discord/voice_state.py

```diff
@@ -0,0 +1,601 @@
+        self._end.clear()
+        self._running.set()
+        try:
+            self._do_run()
+        except Exception:
+            _log.exception('Error in %s', self)
+        finally:
+            self.stop()
+            self._running.clear()
+            self._callbacks.clear()
+
+    def _do_run(self) -> None:
+        while not self._end.is_set():
+            if not self._running.is_set():
+                self._running.wait()
+                continue
+
+            # Since this socket is a non blocking socket, select has to be used to wait on it for reading.
+            try:
+                readable, _, _ = select.select([self.state.socket], [], [], 30)
+            except (ValueError, TypeError):
+                # The socket is either closed or doesn't exist at the moment
+                continue
+
+            if not readable:
+                continue
+
+            try:
+                data = self.state.socket.recv(2048)
+            except OSError:
+                _log.debug('Error reading from socket in %s, this should be safe to ignore', self, exc_info=True)
+            else:
+                for cb in self._callbacks:
+                    try:
+                        cb(data)
+                    except Exception:
+                        _log.exception('Error calling %s in %s', cb, self)
+
+
+class ConnectionFlowState(Enum):
+    """Enum representing voice connection flow state."""
+
+    # fmt: off
+    disconnected            = 0
+    set_guild_voice_state   = 1
+    got_voice_state_update  = 2
+    got_voice_server_update = 3
+    got_both_voice_updates  = 4
+    websocket_connected     = 5
+    got_websocket_ready     = 6
+    got_ip_discovery        = 7
+    connected               = 8
+    # fmt: on
+
+
+class VoiceConnectionState:
+    """Represents the internal state of a voice connection."""
+
+    def __init__(self, voice_client: VoiceClient, *, hook: Optional[WebsocketHook] = None) -> None:
+        self.voice_client = voice_client
+        self.hook = hook
+
+        self.timeout: float = 30.0
+        self.reconnect: bool = True
+        self.self_deaf: bool = False
+        self.self_mute: bool = False
+        self.token: Optional[str] = None
+        self.session_id: Optional[str] = None
+        self.endpoint: Optional[str] = None
+        self.endpoint_ip: Optional[str] = None
+        self.server_id: Optional[int] = None
+        self.ip: Optional[str] = None
+        self.port: Optional[int] = None
+        self.voice_port: Optional[int] = None
+        # TODO: These three MISSINGs are exposed as VoiceClient properties
+        #       I could type ignore them or just leave them as MISSING
+        self.secret_key: List[int] = MISSING
+        self.ssrc: int = MISSING
+        self.mode: SupportedModes = MISSING
+
+        # TODO: These two cause a lot of issues if they're Optional
+        #       Even if they really are optional, access should be already restricted based on state

```

Comment:
I'd say that's kind of the point but okay.

Reply:
Ok the count is 7 `type: ignore`s, 2 `is None` checks, and 1 check I don't even really know how to handle in a sensible way. Here's the diff: https://gist.github.com/imayhaveborkedit/445f7d3a765cf48270000d49f3d5ba79